Source/WebCore/css/parser/CSSPropertyParser.cpp

```diff
@@ -702,70 +702,77 @@ static RefPtr<CSSValue> consumeFontVariantEastAsian(CSSParserTokenRange& range)
     if (range.peek().id() == CSSValueNormal)
         return consumeIdent(range);
     
-    RefPtr<CSSValueList> values = CSSValueList::createSpaceSeparated();
     std::optional<FontVariantEastAsianVariant> variant;
     std::optional<FontVariantEastAsianWidth> width;
     std::optional<FontVariantEastAsianRuby> ruby;
     
-    while (!range.atEnd()) {
-        if (range.peek().type() != IdentToken)
-            return nullptr;
+    auto parseSomethingWithoutError = [&range, &variant, &width, &ruby] () {
+        bool hasParsedSomething = false;
 
-        auto id = range.peek().id();
+        for (;;) {
+            if (range.peek().type() != IdentToken)
+                return hasParsedSomething;
+
+            auto id = range.peek().id();
+            switch (id) {
+            case CSSValueJis78:
+                if (variant)
+                    return false;
+                variant = FontVariantEastAsianVariant::Jis78;
+                break;
+            case CSSValueJis83:
+                if (variant)
+                    return false;
+                variant = FontVariantEastAsianVariant::Jis83;
+                break;
+            case CSSValueJis90:
+                if (variant)
+                    return false;
+                variant = FontVariantEastAsianVariant::Jis90;
+                break;
+            case CSSValueJis04:
+                if (variant)
+                    return false;
+                variant = FontVariantEastAsianVariant::Jis04;
+                break;
+            case CSSValueSimplified:
+                if (variant)
+                    return false;
+                variant = FontVariantEastAsianVariant::Simplified;
+                break;
+            case CSSValueTraditional:
+                if (variant)
+                    return false;
+                variant = FontVariantEastAsianVariant::Traditional;
+                break;
+            case CSSValueFullWidth:
+                if (width)
+                    return false;
+                width = FontVariantEastAsianWidth::Full;
+                break;
+            case CSSValueProportionalWidth:
+                if (width)
+                    return false;
+                width = FontVariantEastAsianWidth::Proportional;
+                break;
+            case CSSValueRuby:
+                if (ruby)
+                    return false;
+                ruby = FontVariantEastAsianRuby::Yes;
+                break;
+            default:
+                return hasParsedSomething;
```

Comment:
My worry is that your logic may allow `font-variant-east-asian: full-width garbage` (though I haven't tested)

The test for invalid values doesn't cover that case (only covers duplicated variant/ruby/width or normal followed by smtg incorrect): https://searchfox.org/wubkat/source/LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-variant-east-asian-invalid.html

Might be nice to test `font-variant-east-asian: full-width garbage` (where garbage is neither variant/ruby/width), or something like  `font-variant-east-asian: full-width none`, and add these to the WPT (I can help with exporting).

Reply:
It works. The range is always checked by the caller function (`parseSingleValue` or `parseShorthandXXX`) which has more context about "should it be the end of a range or not?".

I've added tests.